opentelemetry-sdk/src/opentelemetry/sdk/_metrics/view.py

```diff
@@ -0,0 +1,154 @@
+                the instrument will be used.
+
+            attribute_keys: This is a metric stream customizing attribute: this
+                is a set of attribute keys. If not `None` then only the
+                measurement attributes that are in `attribute_keys` will be
+                used to identify the metric stream.
+
+            aggregation: This is a metric stream customizing attribute: every
+                metric stream has an aggregation instance, this is the class of
+                aggregation this instance will be. If `None` the default
+                aggregation class of the instrument will be used.
+        """
+
+        if (
+            instrument_type
+            is instrument_name
+            is meter_name
+            is meter_version
+            is meter_schema_url
+            is None
+        ):
+            raise Exception(
+                "Some instrument selection "
+                "criteria must be provided for View {name}"
+            )
+
+        self._name = name
+        self._instrument_type = instrument_type
+        self._instrument_name = instrument_name
+        self._meter_name = meter_name
+        self._meter_version = meter_version
+        self._meter_schema_url = meter_schema_url
+
+        self._description = description
+        self._attribute_keys = attribute_keys
+        self._aggregation = aggregation
+
+        if self._name is not None:
+            self._previously_matched_instrumentation_infos = set()
+
+    # pylint: disable=too-many-return-statements
+    # pylint: disable=too-many-branches
+    def _match(self, instrument: Instrument) -> bool:
+
+        if self._instrument_type is not None:
+            if not isinstance(instrument, self._instrument_type):
+                return False
+
+        if self._instrument_name is not None:
+            if not fnmatch(instrument.name, self._instrument_name):
+                return False
+
+        if self._meter_name is not None:
+            if not fnmatch(
+                instrument.instrumentation_info.name, self._meter_name
+            ):
+                return False
+
+        if self._meter_version is not None:
+            if not fnmatch(
+                instrument.instrumentation_info.version, self._meter_version
+            ):
+                return False
+
+        if self._meter_schema_url is not None:
+            if not fnmatch(
+                instrument.instrumentation_info.schema_url,
+                self._meter_schema_url,
+            ):
+                return False
+
+        if self._name is not None:
+            if instrument.instrumentation_info not in (
+                self._previously_matched_instrumentation_infos
+            ):
+                self._previously_matched_instrumentation_infos.add(
+                    instrument.instrumentation_info
+                )
+            else:
+                return False
```

Comment:
I have concerns with this:
1. It's not idempotent i.e. `view.match(counter1)` will return True the first time and False subsequently. Callers of `_match()` probably expect it to be stateless and just say if the instrument matches or not. This is problematic since we do the matching for each MetricReaderStorage
2. Because the first matched instrument wins, a program can output different metrics each time it's restarted. That can be a Heisenbug for users.
3. Josh's most recent PR updated the spec to specify [what producers should do](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#opentelemetry-protocol-data-model-producer-recommendations)
   > Generally, for potential conflicts involving an identifying property (i.e., all properties except description), the producer SHOULD inform the user of a semantic error and pass through conflicting data.
   
   I don't think we should address that in this PR, because it will be a lot of work. It needs to be implemented outside of the View class regardless

Reply:
I have similar concerns, a view with a name should only match one instrument per meter, but which one?

Asked that [here](https://github.com/open-telemetry/opentelemetry-specification/pull/2372#discussion_r813394956).

Reply:
I have removed the stateful behavior of the view and make it raise an exception if it is declared with a name and a wildcard in `instrument_name`.